ci/pingcap_tidb_operator_build_kind.groovy

```diff
@@ -47,104 +134,61 @@ def call(BUILD_BRANCH, CREDENTIALS_ID, CODECOV_CREDENTIALS_ID) {
-					fi
-					go version
-					if ! command -v docker &>/dev/null; then
-						echo "error: docker is required"
-						exit 1
-					fi
-					docker version
-					echo "==== End Build Environment ===="
-					declare -A clusters
-					clusters['kind']='127.0.0.1:5000'
-					clusters['kind2']='127.0.0.2:5000'
-					clusters['kind3']='127.0.0.3:5000'
-					clusters['kind4']='127.0.0.4:5000'
-					while true
-					do
-						for cluster in \${!clusters[*]}; do
-							lockfile=/k8s-locks/\$cluster
-							if [ -e \$lockfile ]; then
-								prId=`cat \$lockfile`
-								if [ "${ghprbPullId}" == "\$prId" ]; then
-									clusterName=\$cluster
-									echo "####### branch ${ghprbPullId}/${BUILD_BRANCH} get env:\$cluster ########"
-									echo "####### if you want to debug,please exec the command: export KUBECONFIG=`/root/go/bin/kind get kubeconfig-path --name=\$cluster ` in jenkins slave node ########"
-										break 2
-								fi
-							fi
-						done
-
-						for cluster in \$(shuf -e \${!clusters[*]}); do
-							lockfile=/k8s-locks/\$cluster
-							if [ ! -e \$lockfile ]; then
-									touch \$lockfile
-									echo ${ghprbPullId} > \$lockfile
-									clusterName=\$cluster
-									echo "####### branch ${ghprbPullId}/${BUILD_BRANCH} get env:\$cluster #######"
-									echo "####### if you want to debug,please exec the command: export KUBECONFIG=`/root/go/bin/kind get kubeconfig-path --name=\$cluster ` in jenkins slave node ########"
-									break 2
+					stage("Build and Test") {
+						ansiColor('xterm') {
+							sh """
+							make
+							make e2e-build
+							if [ ${BUILD_BRANCH} == "master" ]
+							then
+								make test GO_COVER=y
+								curl -s https://codecov.io/bash | bash -s - -t ${CODECOV_TOKEN} || echo 'Codecov did not collect coverage reports'
 							else
-									prId=`cat \$lockfile`
-									echo "env:\$cluster have be used by PR: \$prId"
+								make test
 							fi
-						done
-						echo "sleep 30 second and then retry"
-						sleep 30
-					done
-					trap " echo '############ end of e2e test running on the cluster: \$clusterName ############'; rm -f /k8s-locks/\$clusterName " INT TERM EXIT
-
-					export KUBECONFIG=`/root/go/bin/kind get kubeconfig-path --name="\$clusterName"`
-					DOCKER_REGISTRY=\${clusters[\$clusterName]} IMAGE_TAG=${GITHASH} SKIP_BUILD=y GINKGO_NODES=8 make e2e
-					"""
+							"""
+						}
 					}
+
+					stash excludes: "vendor/**,deploy/**", name: "tidb-operator"
 				}
+			}
+		}
 
+        def builds = [:]
+        builds["E2E v1.12.10"] = {
+			build("IMAGE_TAG=${GITHASH} SKIP_BUILD=y GINKGO_NODES=8 KUBE_VERSION=v1.12.10 make e2e")
+        }
+        builds["E2E v1.16.3"] = {
+			build("IMAGE_TAG=${GITHASH} SKIP_BUILD=y GINKGO_NODES=8 KUBE_VERSION=v1.16.3 make e2e")
+        }
+        builds.failFast = false
+        parallel builds
+
+		// we requires ~/bin/config.cfg, filemgr-linux64 utilities on k8s-kind node
+		// TODO make it possible to run on any node
```

Comment:
Maybe we can upload the binary and chart into internal file server and do the publish in a separate jenkins job which downloads the file from file server and publish to qiniu and other public registries.